src/Pirouette/Transformations/Prenex.hs

```diff
@@ -0,0 +1,163 @@
+--   /\ a : Type . /\ b : Type . \ x : a . \ y : b . ...
+--
+-- This makes it easier for later transformations to apply
+-- on type abstractions. In fact, it's common to just bail
+-- out on non-prenex types in other transformations, because
+-- those cases are quite rare and difficult to handle
+-- (usually involving some impredicative types.)
+prenex :: 
+  PrtUnorderedDefs lang ->
+  PrtUnorderedDefs lang
+prenex PrtUnorderedDefs { prtUODecls, prtUOMainTerm } = do
+  -- this is done in two steps
+  -- step 1. prenexify the types and lambdas
+  let prenexDecls = Map.map prenexDefinitionLambdas prtUODecls
+  -- step 2. prenixify the bodies
+  PrtUnorderedDefs {
+    prtUODecls = Map.map (prenexDefinitionBody prenexDecls) prenexDecls,
+    prtUOMainTerm = prenexExpr prenexDecls prtUOMainTerm
+  }
+
+-- | Update the type and initial lambdas
+-- from a definition. This *must* be followed
+-- by 'prenexDefinitionBody' at a later point
+-- to make everything correct.
+prenexDefinitionLambdas ::
+  Definition lang ->
+  Definition lang
+prenexDefinitionLambdas = onlyOnFunDef $ \FunDef { .. } ->
+  let (newTy, newBody) = switchLambdas funTy funBody
+  in FunDef { funIsRec, funTy = newTy, funBody = newBody }
+
+-- invariant: in the result type all /\ appear before all \
+switchLambdas :: Type lang -> Term lang -> (Type lang, Term lang)
+-- this case pushes lambdas inwards
+-- \ x . /\ a . ty --> /\ a . \x . ty
+switchLambdas (SystemF.TyFun ty1 (SystemF.TyAll annTy kindTy restOfType)) 
+              (SystemF.Lam arg1 tyTm1 (SystemF.Abs annTm kindTm body))
+  = let -- since we are going to have an additional /\ on front,
+        -- we need to shift the bound variables by 1
+        shiftedTy1 = shiftTyBy1 0 ty1
+        shiftedTyTm1 = shiftTyBy1 0 tyTm1
+        (newTy, newBody) = switchLambdas (SystemF.TyFun shiftedTy1 restOfType)
+                           (SystemF.Lam arg1 shiftedTyTm1 body)
+    in (SystemF.TyAll annTy kindTy newTy, SystemF.Abs annTm kindTm newBody)
+-- case of small lambda, we just need to keep the invariant (see *)
+switchLambdas (SystemF.TyFun ty1 ty2) (SystemF.Lam arg1 tyTm1 body)
+  = let (newTy, newBody) = switchLambdas ty2 body
+        newTy' = SystemF.TyFun ty1 newTy
+        newBody' = SystemF.Lam arg1 tyTm1 newBody
+    in case newTy of
+         -- (*) when we have a /\ on the top we are breaking the invariant
+         -- so run this again until everything is fine
+         SystemF.TyAll {} -> switchLambdas newTy' newBody'
+         _ -> (newTy', newBody')
+-- if we have a big lambda, just leave it
+switchLambdas (SystemF.TyAll annTy kindTy restOfType)
+              (SystemF.Abs annTm kindTm body)
+  = let (newTy, newBody) = switchLambdas restOfType body
+    in (SystemF.TyAll annTy kindTy newTy, SystemF.Abs annTm kindTm newBody)
+switchLambdas otherTy otherTm = (otherTy, otherTm)
+
+-- | Shift all the bound type variables,
+-- starting from 'from' upwards.
+shiftTyBy1 :: Integer -> Type lang -> Type lang
```

Comment:
Isn't this just [`shiftCutoff from 1`](https://github.com/tweag/pirouette/blob/main/src/Pirouette/Term/Syntax/Subst.hs#L58)?

Reply:
Not really, looking at the definition of that function:

```haskell
shiftCutoff cutoff k = subst $ foldr (\_ r -> Nothing :< r) (Inc k) [0 .. cutoff - 1]
```

There you need to say (if I understand correctly), until which de Bruijn index you want to shift, so it's an upper bound in the index. The function I've implemented shifts from that de Bruijn index to infinitive, so it works as a lower bound.

Reply:
Unrolling `shiftCutoff 2 4` gives us `subst (Nothing :< Nothing :< Inc 4)`, which means: "leave indexes 0 and 1 untouched and increment anything else by 4"

Reply:
And, while at it, `shiftCutoff` should be called `shiftFrom` now that I think about it

Reply:
I've tested replacing `shiftTyBy1` with `shift` from the `Subst` module, but it doesn't give the right results.

Reply:
Resolved by #73